descartes_light/include/descartes_light/visibility_control.h

```diff
@@ -0,0 +1,55 @@
+// Copyright 2015 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/* This header must be included by all descartes headers which declare symbols
+ * which are defined in the descartes library. When not building the descartes
+ * library, i.e. when using the headers in other package's code, the contents
+ * of this header change the visibility of certain symbols which the descartes
+ * library cannot have, but the consuming code must have inorder to link.
+ */
+#ifndef DESCARTES_LIGHT_VISIBILITY_CONTROL_H
+#define DESCARTES_LIGHT_VISIBILITY_CONTROL_H
+
+// This logic was borrowed (then namespaced) from the examples on the gcc wiki:
+//     https://gcc.gnu.org/wiki/Visibility
+
+#if defined _WIN32 || defined __CYGWIN__
+  #ifdef __GNUC__
+    #define DESCARTES_EXPORT __attribute__ ((dllexport))
+    #define DESCARTES_IMPORT __attribute__ ((dllimport))
+  #else
+    #define DESCARTES_EXPORT __declspec(dllexport)
+    #define DESCARTES_IMPORT __declspec(dllimport)
+  #endif
+  #ifdef DESCARTES_BUILDING_LIBRARY
+    #define DESCARTES_PUBLIC DESCARTES_EXPORT
+  #else
+    #define DESCARTES_PUBLIC DESCARTES_IMPORT
+  #endif
```

Comment:
@johnwason and @gavanderhoorn Do you know why these are in the visibility control file. I pulled this file from rclcpp and with this add you must add the compiler definition DESCARTES_BUILDING_LIBRARY for it to export versus import. When would you every need the import?

Reply:
The export/import direction is switched using the preprocessor flag. When building the library, the `DESCARTES_PUBLIC` is set to `__declspec(dllexport)` using the `DESCARTES_BUILDING_LIBRARY` flag, which should be set in CMake. This causes the symbols marked with `DESCARTES_PUBLIC` to be exported by the DLL. When using the DLL, the `DESCARTES_PUBLIC` is set to `__declspec(dllimport)`, which will then look for the symbols in the DLL. 

The `.obj` file generated contains these exported symbols. This is different than Linux which contains both the DLL and `.obj` in the same `.so` file.

CMake has a macro that generates the export header file and handles the export/import direction change.

https://cmake.org/cmake/help/v3.14/module/GenerateExportHeader.html

This macro is not the same as `WINDOWS_EXPORT_ALL_SYMBOLS`. `WINDOWS_EXPORT_ALL_SYMBOLS` will automatically export all symbols in the DLL when compiling it. `GenerateExportHeader` will instead generate the file that defines `DESCARTES_PUBLIC` and set the `DESCARTES_BUILDING_LIBRARY` when necessary.

Reply:
@johnwason Thank you for the explaination. Do you see any issues doing the way I have? I assume I need to add DESCARTES_BUILDING_LIBRARY as an option in the cmake and default it to ON?

Reply:
Try this command:

    target_compile_definitions(descartes_light PRIVATE DESCARTES_BUILDING_LIBRARY=1)

Where `descartes_light` is the name of the library.